hack/update-knative-components.go

```diff
@@ -0,0 +1,258 @@
+package main
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	github "github.com/google/go-github/v68/github"
+)
+
+var (
+	// search for these variables
+	knSrvPrefix = "knative_serving_version="
+	knEvtPrefix = "knative_eventing_version="
+	knCtrPrefix = "contour_version="
+
+	file = "hack/component-versions.sh"
+)
+
+// get latest version of owner/repo via GH API
+func getLatestVersion(ctx context.Context, client *github.Client, owner string, repo string) (v string, err error) {
+	fmt.Printf("> get latest '%s/%s'...", owner, repo)
+	rr, res, err := client.Repositories.GetLatestRelease(ctx, owner, repo)
+	if err != nil {
+		err = fmt.Errorf("error: request for latest %s release: %v", owner+"/"+repo, err)
+		return
+	}
+	if res.StatusCode < 200 && res.StatusCode > 299 {
+		err = fmt.Errorf("error: Return status code of request for latest %s release is %d", owner+"/"+repo, res.StatusCode)
+		return
+	}
+	v = *rr.Name
+	if v == "" {
+		return "", fmt.Errorf("internal error: returned latest release name is empty for '%s'", repo)
+	}
+	fmt.Println("done")
+	return v, nil
+}
+
+// Read the file where components versions are located. Fetch their version
+// and return them in 'v1.23.0' format (unquoted).
+func getVersionsFromFile() (srv string, evt string, ctr string, err error) {
```

Comment:
One of the benefits of being in Go now is that we can use builtin structure serialization:

```Go
type Versions struct {
	Eventing string
	Serving  string
	Contour  string
}

func ReadVersions(filename string) (v Versions, err error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return
	}
	err = json.Unmarshal(data, &v)
	return
}

```

versions.json:
```json
{
    "Eventing": "v1.2.3",
    "Serving": "v2.2.3",
    "Contour": "v3.2.3"
}
```